### ISSUE TITLE
Connection between ESP and CAN

### ISSUE THREAD

**Issue by Genesisfm** [Adminius/ESP32-ScanMyTesla#2]
Hello,

I have the follow 2 modules:
Mini Kit Esp32 Module https://nl.aliexpress.com/item/1005007235340350.html?spm=a2g0o.order_list.order_list_main.11.103779d2DyVTS3&gatewayAdapt=glo2nld

Sn65hvd230 Vp230 Can Board  https://nl.aliexpress.com/item/1005006438437964.html?spm=a2g0o.order_list.order_list_main.23.103779d2DyVTS3&gatewayAdapt=glo2nld

And my question is how do i have to connect TX and RX from the Sn65hvd230 to the Mini Kit Esp32?
I tryed port 16 and 17 but that doesnt work :(

Thank you BR Robert

**Comment by Adminius** [comment]
Hey Robert,
the modules should be ok. I'm using same in other projects.

But I'm not sure if the sketch still work in general. Can library und ESP IDF had a lot or rewrites. I have to try it by my self...

**Comment by Genesisfm** [comment]
Thank you for your fast reaction, to be shure i want to check my pinout:
![4000f7bb33b4f3fa619e079742fda1535bab8ba9](https://github.com/user-attachments/assets/2e720f6c-54fd-46e8-b5bc-f6d5f35d164c)
I have to connect my Can interfact to I016 and I017?

**Comment by Adminius** [comment]
Yes, 16 and 17

**Comment by Genesisfm** [comment]
Ok i checked that and it doesn't work, i also removed the termination resistor of 120 ohm without any positive result :(

**Comment by Genesisfm** [comment]
I was wondering if you had managed to test it out yet? Perhaps it is possible to share the old libraries it works with?

**Comment by Adminius** [comment]
Hey Robert,
I tested it today and it still works.
[can_common](https://github.com/collin80/can_common) 0.3.0
[ESP32_CAN](https://github.com/collin80/esp32_can) 0.2.3
esp32 Core 2.0.16

I will try to update to the newest versions and try it again.
My idea was to get rid of extra libs, because now esp32 core provides very good api (that is also now used in collin80's libs)
But it means whole rewrite :/ 

What exactly is not working?

I tried with this modules and it doesn't work: https://www.aliexpress.com/item/1005006209071522.html
But it works fine with this: https://www.aliexpress.com/item/1005006408102971.html


Attached is binary that should work.
[ESP32-ScanMyTesla.ino.esp32.zip](https://github.com/user-attachments/files/16502838/ESP32-ScanMyTesla.ino.esp32.zip)

**Comment by Genesisfm** [comment]
Hi,
I ordered an other can module i think there is something wrong with it.
I was able to make a bluetooth connection but a got  a time out on the canbus 😑
Thank you very much.

**Comment by Genesisfm** [comment]
Hi another question, which esp32 module are you using? Thx

**Comment by Genesisfm** [comment]
Hi, 
I tried another CAN bus module but stil the same problem, what is the issue:
When i connect with Scan my tesla i see:
Creating bluetooth socket...
Connecting to OBD2 adapter ESP-SMT...
Initializing adapter ...
Connecting to car...
Using ST1110 command set
Setting filters...
--- the text above will erased---
Then nothing...

**Comment by krasanen** [comment]
This was working fine with SN65HVD230 + ESP32 DEVKIT V1. Maybe double check that RX/TX or CAN-H/CAN-L are not reversed?

**Comment by Adminius** [comment]
I'm using the same as pn the picture abow: Wemos D1 Mini ESP32

P.S. I rewrote the whole sketch to use native ESP32 TWAI driver wirhout external libraries and testing it since few days. I hope I can release it this week.
Also some optimisations were implemented.

**Comment by Genesisfm** [comment]
I currently have several ESP32s, including the ESP32 DEVKIT V1. I still have to test this but I notice that the hardware does not all match (aliexpress...)
Ps. i also not have the orginal Wemos D1 mini ESP but the D1 mini clone...

**Comment by Adminius** [comment]
all my HW is from Ali and all of it are kind of clones :)

**Comment by koncio18** [comment]
Hello, 
Can I change default 16,17 pins in code ? I don't see any definition for it. 
Should BT in the phone see this device in device list without connecting it to the car?  Or only in MyTeslaScan app ? Mine doesn't..

**Comment by Adminius** [comment]
> Should BT in the phone see this device in device list without connecting it to the car

Yes, if you power ESP32 via USB connection. Got to phones BT settings, look for "ESP-SMT" and connect it. Then go to ScanMyTesla app and choose "ESP-SMT" device in the settings

I've just uploaded new version. 
Tested with the latest ESP32 Arduino Core 3.0.4, no external libraries are needed.

Pin-definion: https://github.com/Adminius/ESP32-ScanMyTesla/blob/main/ESP32-ScanMyTesla.ino#L13-L14

**Comment by Adminius** [comment]
@Genesisfm I hope, you know, that you need a special OBD cable, because OBD port in the footwell doesn't provide any usefull information

I'm using this one in my Model Y
https://www.aliexpress.com/item/1005006022463035.html

**Comment by Adminius** [comment]
Hardware Example:
![image](https://github.com/user-attachments/assets/a34a6a93-cde8-4664-8426-c0f9de5fb3e2)
![image](https://github.com/user-attachments/assets/3fabe11c-3d29-45bd-83e2-30556c4d400f)

Links checked today:
Case: https://www.aliexpress.com/item/1468276483.html
ESP32: https://www.aliexpress.com/item/32816045916.html
Step-Down: https://www.aliexpress.com/item/33049921166.html
Protoboard: https://www.aliexpress.com/item/33043148251.html
Can Module: https://www.aliexpress.com/item/1005005334841319.html

**Comment by koncio18** [comment]
Thanks for update. 
Of course I have all hardware. I just tested new version, but I have some issue with BT, on notebook I see ESP-SMT but on iphone 15 not, any suggestion ? (on old and new version)

**Comment by Adminius** [comment]
Argh, iPhone works only with expencive OBDLink.... thanks to Apple :/

**Comment by Genesisfm** [comment]
> @Genesisfm I hope, you know, that you need a special OBD cable, because OBD port in the footwell doesn't provide any usefull information
> 
> I'm using this one in my Model Y https://www.aliexpress.com/item/1005006022463035.html
It works right know with your new software release and i'm using another ESP32 module.
Thank you for your support!

**Comment by dlarue** [comment]
Just wanted to chime in that the code is working fine with ESP-WROOM-32 and Waveshare SN65HVD230 CAN Board with Tx=17 and Rx=16 with Android. 

I'd found a nice compact Waveshare RS485-CAN board which has builtin 7-36VDC regulator and was going to use it but it's based on ESP32S3 which is BLE and not Bluetooth. So I'll be pulling out a DC-DC buck converter and boxing it all up.
Thanks for the work on this.

**Comment by Adminius** [comment]
Yeah, BLE support would be cool, but it's not ready on ScanMyTesla side yet...

**Comment by dzid26** [comment]
This should work https://www.tindie.com/products/fusion/can32-an-esp32-dev-board-with-can-bus-v21/
The only nit is the use of an external antenna.

**Comment by Adminius** [comment]
> This should work https://www.tindie.com/products/fusion/can32-an-esp32-dev-board-with-can-bus-v21/ The only nit is the use of an external antenna.

Ehm, yeah, it should work, but DC input voltage is max. 14.5V only, newer Teslas (cince ~2020) are at 15.5V, so you still need DC-DC stepdown or 5V USB cable. For that price...

**Comment by dzid26** [comment]
I asked the author about the voltage limitation. He said the board uses TPS62160DSGR, which has a maximum of 17V.   
I see he updated the tindie site with max 16V now. - Perhaps limited by typical PTC fuse and tantalum capacitor voltage ratings.  So it should work in the car.

I will just drop few other finds:
https://lilygo.cc/products/t-can485
~~[https://www.mouser.com/ProductDetail/MeatPi/WICAN-OBD-C3](https://eu.mouser.com/ProductDetail/MeatPi/WICAN-OBD-C3?qs=rQFj71Wb1eVeZx56VOjoUg%3D%3D&srsltid=AfmBOorI1XqZFQi6mePFJUkSTIc5deMTmtalK7drnclt3WeBKKsNV0x8)~~  (looks sweet, but it's BLE)
https://aliexpress.com/item/1005007205527422.html (maybe)
https://store.mrdiy.ca/p/esp32-can-bus-shield/ (obd shield only)
https://github.com/MagnusThome/RejsaCAN-ESP32 (nice project - v2 is BT classic)

**Comment by Adminius** [comment]
Let me put it this way: 16V rated capacitor can theoreticaly work with ~15.5V voltage, but not recomended. I heard about max. ~80% of voltage rating are safe. 25V rating would be better.

In general every board with ESP32 (NOT -C/S/H!!) should work. New -C/S/H Version doesn't support BT Classic = UART via Bluetooth.

**Comment by dzid26** [comment]
My experience is that ~80% headroom is a bit of a myth as components have their own headroom. If we are within the rating and don't expect spikes, we should be fine. Though it probably makes sense for designs that will be produced in millions of units. :)

A see the price wise, some of those product are getting close to the "official" https://sensotech.it/products/obdii-bluetooth-dongle-vgate-for-android-ios-scan-my-tesla-approved-apple-mfi-certified territory - currently at 59.90EUR.

Ah, I see it's the ScanMyTesla app limitation to use Bluetooth classic...
A bit offtopic, I think it would be possible to use ScanMyTesla with ESP32-C/S/H Version - over wifi - by adapting [CANserver V1](https://github.com/joshwardell/CANserver/blob/v1/CANserver/esp32_can/esp32_can.cpp#L5) to it - if someone is really pressed. 

I think [your] bluetooth solution is nicer though, as it won't mess with phone internet connection. 
Thanks

**Comment by dlarue** [comment]
Since the subjects are power input(DC-DC), ESP32-C/S/H, Bluetooth and CAN I just wanted to add that a happy in-between of a setup with ESP32+CAN+DC-DC would be something like the WaveShare RS485-CAN which has in input voltage of 7-36VDC, CAN and screw terminals for all the connections. Missing is BT UART which can be added with a small BT board and since there's no physical external connections to the BT board it could just be shink-wrapped to the OBD-II wiring or possibly packed inside its case. 
https://www.waveshare.com/esp32-s3-rs485-can.htm

**Comment by Adminius** [comment]
Or just buy cheap components from Ali, 5 min of soldering and you are done :)

EDIT: I'm not recomending using Canserver V1. V2 ist way better, but closed source.
Still, if you use ESP32 with "standard" MOSI/MIS/SCK Pins and CS=GPIO5 for SD card, 16+17 for CAN than CANServer V2 will work on you own hardware ;)

**Comment by dlarue** [comment]
> Or just buy cheap components from Ali, 5 min of soldering and you are done :)

Yes, been there, done that but "and you are done" isn't correct. You still need a case and on the Model Y there's no space in the rear of the center console where the OBD-II port has been exposed with a splitter which itself almost doesn't fit.
The bunch-o-parts method worked great for me to get my data out of the battery before it was replaced but now want to do some regular recording so a more permanent setup is desired and the smaller the better because it might get tucked under a seat or even slipped into a seatback pocket.

**Comment by Adminius** [comment]
this thing "flies arond" between middle console and driver seat rail since 3 years, No problem at all.
https://github.com/Adminius/ESP32-ScanMyTesla/issues/2#issuecomment-2351022973

buy a cheap CAN OBD cable, like https://aliexpress.com/item/1005008225592747.html, throw away the OBD bulky connector and solder directly to "sandwich" style esp32, as on my picture.
add some duct tape and hide it completly. nobody needs OBD plug/socket.

**Comment by Adminius** [comment]
alternatively:
Buy this "highland" cable for, there are two busses available.
https://tessi-supply.com/products/sexy-buttons-v2-passend-fur-deinen-tesla-model-3-y

I'm working on "easy to use" hardware to connect both busses and maybe we can get also second bus into ScanMyTesla. Let's see, but it take time, just started

**Comment by dlarue** [comment]
Dude! There's a feak'n WeMos power shield?  Fitting that inside an OBD-II shell if fantastic. The OBD-II cable/connector I'd purchased is sealed so I had to hang my parts off the cable end. The power shield and cable/connector will be ordered today.
Definitely like the unit in the connector setup. Thx.

**Comment by Adminius** [comment]
> There's a feak'n WeMos power shield?

Yes, but with desoldered black 5.5x2.1 socket

SD Card shield ist also there 🤣

**Comment by dlarue** [comment]
Piece of cake, I used to be an Electronics Tech decades ago, soldering is in my blood and I have a bag of SD card holders from a ocean tide charting project(also ESP32 based). Since you mentioned it, I'm guessing the firmware logs locally if enabled so that's a cool feature and will look into if it fits in the connector case.  Do I see the side of the uSD card holder in the opposite side of the ESP32 and on the under the prototype shield?

**Comment by Adminius** [comment]
> under the prototype shield

Yes,on  wemos proto shield (or maybe not, can't rember any more, maybe proto + generic micro sd card)

I used it for CAN Server firmware. My Firmware doesn't use it
